packages/SwingSet/src/vats/comms/delivery.js

```diff
@@ -132,19 +137,29 @@ export function makeDeliveryKit(state, syscall, transmit, clistKit, stateKit) {
       // changed to assert here that the result parameter is null or undefined.
       syscall.resolve([[result, false, UNDEFINED]]);
     }
-    // The message is preceded by an optional sequence number:
-    // `$seqnum:$actualMessage` or `:actualMessage`
-    const colon = message.indexOf(':');
-    assert(colon >= 0, X`received message ${message} lacks seqNum delimiter`);
-    const seqNum = message.substring(0, colon);
+    // The message is preceded by an optional sequence number followed by the
+    // sequence number that the remote end had most recently received from us as
+    // of sending the message:
+    // `$seqnum:$ackSeqNum:$actualMessage` or `:$ackSeqNum:$actualMessage`
+    const delim1 = message.indexOf(':');
+    assert(delim1 >= 0, X`received message ${message} lacks seqNum delimiter`);
+    const seqNum = message.substring(0, delim1);
     const remote = getRemote(state, remoteID);
+    remote.lastReceivedSeqNum += 1;
     assert(
-      seqNum === '' || seqNum === `${remote.nextExpectedRecvSeqNum}`,
+      seqNum === '' || seqNum === `${remote.lastReceivedSeqNum}`,
       X`unexpected recv seqNum ${seqNum}`,
     );
-    remote.nextExpectedRecvSeqNum += 1;
 
-    const msgBody = message.substring(colon + 1);
+    const delim2 = message.indexOf(':', delim1 + 1);
+    assert(
+      delim2 >= 0,
+      X`received message ${message} lacks ackSeqNum delimiter`,
+    );
+    const ackSeqNum = message.substring(delim1 + 1, delim2);
+    handleAckFromRemote(remoteID, ackSeqNum);
```

Comment:
Do we want to pass the `ackSeqNum` as a string here? Or would it be better as a Number/Nat/BigInt?

Also, we talked about keeping this field smaller by sending a delta instead of a full (eternally-increasing) integer.. what are your current thoughts on that? I imagine that debugging will be easier if we make it a full acknum, and also include the (optional) sequnum, so we can compare incoming/outgoing messages by eye.

We could maybe augment the protocol with a `+N` syntax for this field to switch to the delta approach later without causing a compatibility break. Basically if `ackSeqNum` (the string) begins with `+` then convert the rest to an integer, add it a previously-remembered acknum, and deliver it to `handleAckFromRemote`. If not, convert the whole thing to an integer, *update* the remembered acknum, and deliver it.

As iit stands, we might want an assertion to check that the acknums are never decreasing. Maybe.. not sure what value that would add.

Reply:
Hmm.  Yes, I think we want to be converting to number deliberately and carefully.  This use of automagic invisible conversion slipped through.

I was envisioning the ack as a delta as a separate optimization once we are convinced this is working.  Debugging/testing will be easier with the full sequence number for now.  I think syntactically distinguishing the delta is maybe too suspenders-and-belt even for me, unless you are thinking we'd maintain a way to turn it on and off for debugging.